server/src/main/java/org/candlepin/controller/ActivationKeyController.java

```diff
@@ -0,0 +1,246 @@
+     */
+    private ActivationKey fetchActivationKey(String keyId) {
+        if (keyId == null || keyId.isEmpty()) {
+            throw new IllegalArgumentException(i18n.tr("activation key ID is null or empty"));
+        }
+
+        ActivationKey key = this.activationKeyCurator.secureGet(keyId);
+
+        if (key == null) {
+            throw new NotFoundException(i18n.tr("ActivationKey with id {0} could not be found.", keyId));
+        }
+
+        return key;
+    }
+
+    public ActivationKey getActivationKey(String activationKeyId) {
+        return this.fetchActivationKey(activationKeyId);
+    }
+
+    public Set<ActivationKeyPool> getActivationKeyPools(String activationKeyId) {
+        return this.fetchActivationKey(activationKeyId)
+            .getPools();
+    }
+
+    public ActivationKey updateActivationKey(String activationKeyId, ActivationKeyDTO update) {
+        ActivationKey toUpdate = this.fetchActivationKey(activationKeyId);
+
+        if (update.getName() != null) {
+            Matcher keyMatcher = AK_CHAR_FILTER.matcher(update.getName());
+
+            if (!keyMatcher.matches()) {
+                throw new IllegalArgumentException(
+                    i18n.tr("The activation key name \"{0}\" must be alphanumeric or " +
+                        "include the characters \"-\" or \"_\"", update.getName()));
+            }
+
+            toUpdate.setName(update.getName());
+        }
+
+        String serviceLevel = update.getServiceLevel();
+        if (serviceLevel != null) {
+            serviceLevelValidator.validate(toUpdate.getOwner().getId(), serviceLevel);
+            toUpdate.setServiceLevel(serviceLevel);
+        }
+
+        if (update.getReleaseVersion() != null) {
+            toUpdate.setReleaseVer(new Release(update.getReleaseVersion()));
+        }
+
+        if (update.getDescription() != null) {
+            toUpdate.setDescription(update.getDescription());
+        }
+
+        if (update.getUsage() != null) {
+            toUpdate.setUsage(update.getUsage());
+        }
+
+        if (update.getRole() != null) {
+            toUpdate.setRole(update.getRole());
+        }
+
+        if (update.getAddOns() != null) {
+            Set<String> addOns = new HashSet<>(update.getAddOns());
+            toUpdate.setAddOns(addOns);
+        }
+
+        if (update.isAutoAttach() != null) {
+            toUpdate.setAutoAttach(update.isAutoAttach());
+        }
+        return activationKeyCurator.merge(toUpdate);
+    }
+
+    public ActivationKey addPoolToKey(String activationKeyId, String poolId, Long quantity) {
+        ActivationKey key = this.fetchActivationKey(activationKeyId);
+        Pool pool = findPool(poolId);
+
+        // Throws a RuleValidationException if adding pool to key is a bad idea
+        String message = activationKeyRules.validatePoolForActivationKey(key, pool, quantity);
+        if (message != null) {
+            throw new RuleValidationException(message);
+        }
```

Comment:
The throwing of such an exception probably should be part of the rules API itself (thrown by `validatePoolForActivationKey`).

Reply:
Do you want it just declared as throws or do you think it should be checked exception?